### DISCUSSION TITLE
How to form and decorate groups of edges?

### DISCUSSION THREAD

**Discussion by mrcalvin** [eclipse-glsp/glsp#1307] [General]
I am aware of how to add decorators to single edges (like in https://github.com/eclipse-glsp/glsp/discussions/1113). However, I have the following requirement (to realise one type of [feature diagrams](https://www.featureide.de/#screenshots_content)):

<img width="211" alt="image" src="https://github.com/eclipse-glsp/glsp/assets/442057/0b485729-c1ec-41d5-bdd6-5c73b9b9b1c9">

Groups of two or more edges (at one end) need to be grouped by a filled (or, empty) circular section (with the section be delimited by the two outer edges of the group).

I would appreciate any hints! Any implementations available?

**Comment by planger** [comment]
For this requirement, there isn't an off-the-shelf solution readily available to my knowledge. However, I think this can be done with a customization of the view implementation or a node post-processor, adding some custom SVG in the rendering process.

Here are two approaches you might consider:

1. **Custom Edge View Implementation**: This involves registering a custom edge view where you can add a condition in the `renderLine` method. This condition checks if the edge requires a circular sector and if it is e.g. the left-most among the connected edges. If true, locate the right-most edge and add a circle in SVG at the correct position, clipping it to the angles defined by the left-most and right-most edges' first segments. A somewhat similar strategy was used for drawing edge intersections in the Sprotty framework (more details [here](https://github.com/eclipse-sprotty/sprotty/blob/master/packages/sprotty/src/graph/views.tsx#L137)).

2. **VNode Postprocessor**: Another option is to register a VNode postprocessor, which is invoked after the entire SVG has been created. Such postprocessors can then afterwards change or add SVG nodes based on the virtual DOM representing the SVG. Thus, you could identify such cases and add the necessary SVG accordingly. More on VNode postprocessors can be found [here](https://github.com/eclipse-sprotty/sprotty/blob/master/packages/sprotty/src/base/views/vnode-postprocessor.ts).

I'd recommend starting with the first approach (Edge View implementation). To register a custom edge view, you can use the following code snippet:

```ts
configureModelElement(context, 'edge', SEdgeImpl, YourEdgeViewImpl);
```

In your `YourEdgeViewImpl`'s `renderLine` method, you could check if the current edge is the left-most among all outgoing edges from its source. If it is, then proceed with obtaining the relevant other edges (e.g. via navigating to the source and then finding all outgoing edges), obtain their first segments, etc. There is hopefully some useful code in this direction in the [implementation here](https://github.com/eclipse-sprotty/sprotty/blob/master/packages/sprotty/src/graph/views.tsx#L137).

Hope this helps, and I'm keen to see how your implementation turns out!

**Comment by mrcalvin** [comment]
Thx @planger for the pointers! I will investigate a custom `EdgeView`, in fact, there is already one in place to render (filled, empty) ball end decorators. I will report back.
